models.py

```diff
@@ -177,54 +179,218 @@ def __str__(self):
+            )
+        )
+
+        explicit_links = filter(
+            lambda link: link is not None,
+            map(
+                self.controller.napps[('kytos', 'topology')].links.get,
+                window.links
             )
-            self.controller.buffers.app.put(event)
-        if window.interfaces:
-            event = KytosEvent(
-                name=f'kytos/maintenance.{operation}_interface',
-                content={'unis': window.interfaces}
+        )
+
+        tot_links = list(chain(implicit_links, explicit_links))
+
+        affected_switch_ids = list(set(map(
+            lambda switch: switch.id,
+            filter(
+                self.switch_not_in_maintenance,
+                tot_switches
+            )
+        )))
+
+        affected_interface_ids = list(set(map(
+            lambda interface: interface.id,
+            filter(
+                self.interface_not_in_maintenance,
+                tot_interfaces
             )
-            self.controller.buffers.app.put(event)
-        if window.links:
-            event = KytosEvent(
-                name=f'kytos/maintenance.{operation}_link',
-                content={'links': window.links}
+        )))
+
+        affected_link_ids = list(set(map(
+            lambda link: link.id,
+            filter(
+                self.link_not_in_maintenance,
+                tot_links
             )
-            self.controller.buffers.app.put(event)
+        )))
+
+        return {
+            'switches': affected_switch_ids,
+            'interfaces': affected_interface_ids,
+            'links': affected_link_ids,
+        }
 
     def start_mw(self, window: MaintenanceWindow):
         """Actions taken when a maintenance window starts."""
-        items = [*window.switches, *window.links, *window.interfaces]
-        self.maintenance_devices.update(items)
-        self._maintenance_event(window, 'start')
+        affected_ids = self._get_affected_ids(window)
+
+        self.maintenance_switches.update(window.switches)
+        self.maintenance_interfaces.update(window.interfaces)
+        self.maintenance_links.update(window.links)
+
+        self._maintenance_event(
+            affected_ids,
+            'start'
+        )
 
     def end_mw(self, window: MaintenanceWindow):
         """Actions taken when a maintenance window finishes."""
-        items = [*window.switches, *window.links, *window.interfaces]
-        self.maintenance_devices.subtract(items)
-        self._maintenance_event(window, 'end')
 
-    def dev_in_maintenance(self, dev):
+        self.maintenance_switches.subtract(window.switches)
+        self.maintenance_interfaces.subtract(window.interfaces)
+        self.maintenance_links.subtract(window.links)
+
+        affected_ids = self._get_affected_ids(window)
+
+        self._maintenance_event(
+            affected_ids,
+            'end'
+        )
+
+    def switch_not_in_maintenance(self, dev: Switch):
```

Comment:
It'd be great if you could also add a type hint for the return type, we aren't linting for this yet, but especially for brand new code it'd be great to have it with typing hints if possible. Same comment for the methods below.

Reply:
Added in the type hints for return type.